### ISSUE TITLE
Errors within instantiations of templates in system headers not diagnosed

### ISSUE THREAD

**Issue by helixhorned** [clangd/clangd#775]
Hello,

I've been using clangd for a while now and what for me makes coding significantly more convenient is the ability to see accurate compilation errors and warnings inside the editor. So, big thanks to all contributors!

----

There are currently cases where clangd considers a source file to have no errors when invoking compilation of the TU it belongs to would fail. This seems to be the case when the location of the error is inside a system header. The primary example here is an error inside the instantiation of a template deep into the C++ standard library, when however the "logical place" of the error is in user code. I've created a small test suite exposing the issue in some detail:

 https://github.com/helixhorned/clangd-error-in-template-test/commit/6bb1fc62f8a5d14e945284a3fddd34e06af47072

It contains three test cases all exercising a template `DoEmplace<>(char key, Args&&...)` wrapping C++17's `std::map<>::try_emplace()` and allows using a system-installed `clang++`/`clangd`, or a self-built one [1]. In order from simplest to most complex, these are called:

- `ok`: TU is well-formed
- `ill_ok`: TU is ill-formed and clangd diagnoses this
- `ill_nok`: TU is ill-formed and clangd does not diagnose this (the subject of this issue)

[1] The behavior described here can still be observed as of commit https://github.com/llvm/llvm-project/commit/4624412367f9b591fe90ecec8feed7209cd222ac.

As already mentioned, what is specific to `ill_nok` is that the location of the actual **error** is inside C++ library code. On the user code site, we merely have a **note**:

```c++
    /*
     | In file included from ./test.cpp:1:
     | In file included from (...)/include/c++/10/map:60:
     |  [snip]
 lib | .../new_allocator.h: error: no matching constructor ...
     | .../alloc_traits.h: note: in instantiation ...
     |  [snip]
user | ./test.cpp: note: in instantiation (...) 'std::map<char, Class>::try_emplace ...' ...
user | ./test.cpp: note: in instantiation (...) 'DoEmplace<int, int>' requested here
     |  [snip]
    */
    DoEmplace('d', 4, 0xBAD);  // too many arguments to any 'Class' ctor
```

My understanding is that because clangd is intended to be used from inside an editor/IDE, its model is to some extent (source-)file-based as opposed to TU-based [2]. Nevertheless, the `ill_ok` case is now [3] constructed in a way that involves a `common.h` containing the definition of `DoEmplace<>()`, and one **does** see an error when looking at `test_ill_ok.cpp` in the editor. So, again, the deciding factor seems to be the fact that with `ill_nok`,

- the error is inside the system library, and
- at the location one is usually [4] interested in, only a note appears.

[2] An interesting question to ponder about is "what happens if the relation is not one-to-one".
[3] (as opposed to the time when the comments of the code snippet have been extracted)
[4] (that is, when not implementing parts of the C++ standard library itself)

----

Would you agree with the classification of this behavior as something unexpected (i.e. a bug)?

If yes, would it be reasonable to have a sort of heuristic [5] when a now-filtered-out diagnostic should instead make it as reply to a query for diagnostics? Or, the other way around: what is the reason that these are (presumably) filtered out now? The presumed notion that diagnostics from within system headers are not of interest to the user? I would agree for **warnings**, but not for errors.

[5] Ideally, it would not be a heuristic but just logic that makes sense in general, of course.

Cheers,
Philipp

----

**Logs** (lightly edited)
[clangd_ill_ok.log](https://github.com/clangd/clangd/files/6472432/clangd_ill_ok.log)
[clangd_ill_nok.log~](https://github.com/clangd/clangd/files/6472433/clangd_ill_nok.log)

**System information**
Output of `clangd --version`:

```
clangd version 13.0.0 (https://github.com/llvm/llvm-project.git 4624412367f9b591fe90ecec8feed7209cd222ac)
```

Editor/LSP plugin: Emacs/eglot (but reproducible stand-alone)
Operating system: Ubuntu 20.10

**Comment by kadircet** [comment]
Thanks for the amazing repros and such. Unfortunately this is intended behavior (or known issue?). It is not because we filter the diagnostics outside the user code but because of a tradeoff we make between correctness and speed.

clangd skips function bodies (most of the time) when parsing headers (anything outside the main file, whether user or system header) to speed things up (definitions aren't really useful apart from diagnosing such failures, and even though i don't remember the exact numbers, the speedup is huge!).

In this case, no function body -> no instantiation of the template -> no diagnostics.

You can observe this more closely with a single file that contains the function bodies:
```
template <typename K, typename V>
struct map {
  template <typename... Args>
  void try_emplace(K, Args &&...args) {
    V v(args...);
  }
};

class Class {
  Class(void *) {}
 public:
  Class(int);
};

template <typename... Args>
void DoEmplace(char key, Args &&...args) {
  map<char, Class> m;
  m.try_emplace(key, args...);
}

inline void ill_nok() {
  DoEmplace('d', 4, 0xBAD);  // too many arguments to any 'Class' ctor
  DoEmplace('e', reinterpret_cast<void *>(0xBAD));  // private ctor
}
```

We should definitely document this better in user visible places. 

Also we actually preserve the function body for make_unique, because it is just a single statement (hence cheap to instantiate and parse) and also very common. I remember we discussed others like make_shared, try_emplace etc. but i think we never got to experiment to see the amount of slowdown.

**Comment by HighCommander4** [comment]
See also previous discussion in #137.

**Comment by helixhorned** [comment]
Thanks for the enlightenment, @kadircet and @HighCommander4!

I was about to write down the suggestions I had in mind in the hope that a solution with only a justifiable performance impact is somewhere out there, but the conclusion of #137 seems very close to a "probably won't address" to me.

Looking at the code, I came up with a workaround. It's all about the "preamble" after all, which as far as I recall (and apparently confirmed by my workaround working) is the longest prefix of preprocessor directives for a given source file. The workaround then is to temporarily place on its first line some placeholder non-preprocessor code like `int qwe;`. This works out of the box for me when the TU for the source file does not include a PCH, may give trouble if it does, and I haven't yet tested it with non-source files.

**Comment by kadircet** [comment]
yes your workaround should work. but the problem is, it throws away all the speed optimizations clangd has. Clangd will basically parse all of the main file contents (everything after the preamble) before answering requests like code completion, signature help, hover etc. and if your main file expands to millions of lines of code after preprocessing (because you now have lots of includes), operations latency will increase drastically.

**Comment by helixhorned** [comment]
> [the workaround] throws away all the speed optimizations clangd has.

I understand this. The workaround is supposed to be used on encountering an error for the duration of fixing it.

> (...) operations latency will increase drastically.

... but will still better than the edit+compile+look-at-error-output cycle one would have to live with otherwise. And error logs of the `ill_nok` kind are long and unwieldy.

----

The more I think about this, the more I'm leaning toward a client-side solution. After all, I see that having anything like the workaround enabled constantly is not an acceptable solution in general. (Errors of this kind are made not too frequently, I would say.)

So, I guess I'm fine with closing this issue. I'll leave that to you though because you might want to keep it open as a reminder for e.g.

> We should definitely document this better in user visible places.

**Comment by kadircet** [comment]
> So, I guess I'm fine with closing this issue. I'll leave that to you though because you might want to keep it open as a reminder for e.g.

Thanks! Yes I've took note about that hence keeping the issue around for now.
